seeker.go

```diff
@@ -0,0 +1,192 @@
+package roundtrip
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	stateOK = iota
+	stateEOF
+	stateError
+	stateClosed
+)
+
+// Seeker is a file download object
+// that supports io.ReadSeeker interface
+type seeker struct {
+	client        *Client
+	endpoint      string
+	fileSize      int64
+	currentReader io.ReadCloser
+	currentOffset int64
+	lastError     error
+	state         int
+}
+
+func (s *seeker) updateState(err error) error {
+	switch s.state {
+	case stateError:
+		return err
+	case stateOK, stateEOF:
+		if err == nil {
+			return nil
+		}
+		if err == io.EOF {
+			s.state = stateEOF
+			s.lastError = err
+			return err
+		}
+		s.state = stateError
+		s.lastError = err
+		return err
+	}
+	return err
+}
+
+func (s *seeker) canRead() error {
+	if s.state == stateOK {
+		return nil
+	}
+	return s.lastError
+}
+
+func (s *seeker) canSeek() error {
+	if s.state == stateOK || s.state == stateEOF {
+		return nil
+	}
+	return s.lastError
+}
+
+func newSeeker(c *Client, endpoint string) (ReadSeekCloser, error) {
+	response, err := c.RoundTrip(func() (*http.Response, error) {
+		req, err := http.NewRequest("HEAD", endpoint, nil)
+		if err != nil {
+			return nil, err
+		}
+		c.addAuth(req)
+		return c.client.Do(req)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	sizeHeader := response.headers.Get("Content-Length")
+	size, err := strconv.ParseInt(sizeHeader, 10, 0)
+	if err != nil {
+		return nil, trace.BadParameter("failed to determine size, Content-Length:%v", sizeHeader)
+	}
+	return &seeker{client: c, endpoint: endpoint, fileSize: size}, nil
+}
+
+func (s *seeker) Read(p []byte) (int, error) {
+	if err := s.canRead(); err != nil {
+		return 0, err
+	}
+
+	err := s.initReader()
+	if err != nil {
+		return 0, trace.Wrap(s.updateState(err))
+	}
+
+	bytesRead, err := s.currentReader.Read(p)
+	s.currentOffset += int64(bytesRead)
+
+	return bytesRead, s.updateState(err)
+}
+
+func (s *seeker) initReader() error {
+	if s.currentReader != nil {
+		return nil
+	}
+	// If the offset is great than or equal to size, return a empty, noop reader.
+	if s.currentOffset >= s.fileSize {
+		return trace.BadParameter("offset(%v) is > size(%v)", s.currentOffset, s.fileSize)
+	}
+
+	req, err := http.NewRequest("GET", s.endpoint, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.client.addAuth(req)
+
+	if s.currentOffset > 0 {
+		// If we are at different offset, issue a range request from there.
+		req.Header.Add("Range", fmt.Sprintf("bytes=%v-/%v", s.currentOffset, s.fileSize))
```

Comment:
to make it clear, does `a-/b` as `Range` specification mean a range from an offset `a` to the end of a total `b`?
I just could not find a formal specification of the `Range` header that mentions that syntax. It does match the `Content-Range` syntax though: https://tools.ietf.org/html/rfc7233 (Appendix D, Collected ABNF) - hence my question.